### DISCUSSION TITLE
iManufactorer, iProfuct and iSerialNumber  show "Error Accessing string" on Linux

### DISCUSSION THREAD

**Discussion by jerome-diver** [pyusb/pyusb#421] [Q&A] [answered]
OS: Linux Fedora-34
for each device, i get only "Error Accessing string" (or nothing if there is nothing)
for iManufactorer, iProfuct and iSerialNumber
But lsusb command show me all of them that PyUSB failed to show.

Is it a real multi-platform able package or just a Windows club specific package ?
How to help to fix this ?

**Comment by jonasmalacofilho** [comment]
> Is it a real multi-platform able package or just a Windows club specific package ?

I'll assume it wasn't your intention, but this reads as condescending (in two separate ways).  Still, please try to avoid it in the future.

---

You have a permission issue.

Unlike on Windows, USB devices are normally _not_ accessible by unprivileged users on Linux, and reading descriptors like `iManufacturer` has to access the device.

**Comment by jerome-diver** [comment]
Sure, as you can said, you made an "intention process" by the way to define it as "condescendant". It was not my intention, but you judge it this way... I'm very sorry for that. Can i ask you to avoid to build intentions process and next, to make me paid for that ? You are free to practice what you want, but i think it is not constructive to imagine intentions like that, recognize it as it is or not, never mind, it is not the subject.

So i asked because i'm under Linux and i can read you would like it to be all platform compatible... but if it can not fully works on any other platform than Windows, why not ? Just assume it. 

Or maybe we can try to find how to fix it for it can also works on Linux.

What make me feel strange about: "Unlike on Windows, USB devices are normally not accessible by unprivileged users on Linux, and reading descriptors like iManufacturer has to access the device." (that is maybe true) is that, again (read it back) "lsusb" command (not in root user) provide Manufacturer information, so then it can be possible to get it out of root user. It is a fact, install any Linux distro on any virtual environment and check it by yourself. But not only "lsusb" command can access USB device without to be run under privileged user... in fact, all the other commands. I'm talking about facts (or i miss-understand your point).

Well... that said, i hope it will not be a kind of problem to just talk about facts and try to find, not an explication about why not to do it, but better: how to do it.
I think it will be far more constructive. Isn't it by the facts ?
Thank you for your intention and again, very sorry if your perception was negative from my bad communication (it should have some cultural difference maybe... let's try to be positive and not imagine things sadly, please).

**Reply by jonasmalacofilho** [reply]
>  What make me feel strange about: "Unlike on Windows, USB devices are normally not accessible by unprivileged users on Linux, and reading descriptors like iManufacturer has to access the device." (that is maybe true) is that, again (read it back) "lsusb" command (not in root user) provide Manufacturer information, so then it can be possible to get it out of root user. It is a fact, install any Linux distro on any virtual environment and check it by yourself.

Nowadays, lsusb reads the information from udev, not directly from the device.

(You can check the latter statement with `LIBUSB_DEBUG=4` and/or using `strace`).

> But not only "lsusb" command can access USB device without to be run under privileged user... in fact, all the other commands. I'm talking about facts (or i miss-understand your point).

Which other programs, exactly?

**Reply by jerome-diver** [reply]
how many other should exist for be considered more than already one ("lsusb") ?
Do you think it would be possible that PyUSB read missing information from udev the same way lsusb is doing ? The idea is to be able to read what is actually missing. And if this is not a good idea, which other idea can fix the problem to be full usable on Linux ?
And why do you consider it to not be a bug ? I think it is not a discussion, but an issue or missing functionality to be compatible really with Linux. Is there any susceptibility with the subject (and if yes, why ? I'm asking because i maybe missing some historic around  ?). It seems strange to me to see this moved there and also rejected so fast.

**Reply by jonasmalacofilho** [reply]
Hey Jerome,

> Do you think it would be possible that PyUSB read missing information from udev the same way lsusb is doing ?

lsusb doesn't necessarily show the same information, and, more importantly, has a totally different purpose.[^1]

[^1]: Nowadays, `lsusb -v` can indeed show the same information by taking it from from udev (from when the kernel probed the device).  But without `-v` (or in older versions), `lsusb` generally shows information _externally_ associated to the device's `idVendor` and `idProduct`, for example from the USB ID Repository, which is in no way what the device itself may report with `iManufacturer` and `iProduct`.  If you need further clarification on this difference, please read the USB specification and take a better look into the USB ID Repository.

Still, it would be technically possible for PyUSB (or LibUSB, which would probably make a bit more sense) to fetch the descriptors from udev.  But that's not actually that valuable when you plan to read/write from/to the device anyway... which I assume is the end goal, otherwise why are you using PyUSB/LibUSB?

Also, adding the capability for PyUSB to automatically ("automagically") fetch these descriptors from udev would make our API specification more complex, possibly confusing: currently, `<Device>.manufacturer`, `<Device>.product` and `Device.serial_number` have very simple specifications, and return the corresponding string descriptors, _as reported by the device._  Adding udev to the mix would require us to specify that, actually, the information may come from udev, and then get into details on when might that happen, what exactly may udev report, and how to bypass udev if one needs to.

---

> The idea is to be able to read what is actually missing. And if this is not a good idea, which other idea can fix the problem to be full usable on Linux ?

It _is_ fully usable on Linux.[^2]

[^2]: In fact, like a lot of open-source projects, our fault lies in being Linux-first, even if our goal is to provide cross-compatibility with Windows and other OSs.  And, personally, I use Linux around 90% of the time (even more so when working on software or hardware).

Either run your program from a user that already has privileges to read/write from that device, or give your user the necessary permissions to do that.  We generally use udev rules for the latter.

Specifically, on systemd systems, we generally use the (not very well documented) `TAG+="uaccess"` mechanism.[^3]  With that mechanism,  each rule looks something like this:

[^3]: https://github.com/systemd/systemd/issues/4288

```
SUBSYSTEMS=="usb", ATTRS{idVendor}=="1e71", ATTRS{idProduct}=="1711", TAG+="uaccess"
```

Alternatively, you can use `MODE` and `GROUP` keys.  This may even be required if your sessions on that system may be remote-only, because `uaccess` only applies to local sessions ("seats").

```
SUBSYSTEMS=="usb", ATTRS{idVendor}=="1e71", ATTRS{idProduct}=="1711", GROUP="plugdev", MODE="0660"
SUBSYSTEMS=="usb", ATTRS{idVendor}=="1e71", ATTRS{idProduct}=="1711", MODE="0666"
```

---

> And why do you consider it to not be a bug ? I think it is not a discussion, but an issue or missing functionality to be compatible really with Linux. Is there any susceptibility with the subject (and if yes, why ? I'm asking because i maybe missing some historic around ?). It seems strange to me to see this moved there and also rejected so fast.

A bug is a specific problem where the implementation deviates from its specification, or where the specification deviates from its goal (or, transitively, when there is no specification and where the implementation deviates from its goal).  None of these apply here (and it already is  "usable"/"compatible with" Linux).

Issues are taken to be more general problems, like requests for added functionality.  Still, in your original message you asked a _question,_ not reported a bug or proposed a new feature... which is why this was moved from issues to Q&As (= "questions & answers").

Also, it wasn't rejected "so fast": it just wasn't _immediately_ accepted, but we're still discussing the subject, aren't we?

**Comment by jerome-diver** [comment]
Good to know... i should investigate when having more time...
I'm using PyUSB for the forst time to try (and will use pyserial to communicate with an Arduino, but still would be happy to use PyUSB to build a kind of dialog box with PySide6 to show USB devices and interfaces connected with full informations.

> A bug is a specific problem where the implementation deviates from its specification, or where the specification deviates from its goal (or, transitively, when there is no specification and where the implementation deviates from its goal). 

This is exactly why i was thinking it is a bug, because there is in the device tree some title: iManufacturer (and all concerned) that exist and print an error message. What is un-logic there ? And if you convince me for the difficulty to embed udev to the mix, you failed to demonstrate that it is not a bug (as long as the iManufacturer and other exist and print an error message). It seems to me more a rhetoric point of view that some arguments that you did exposed (hope that does not offense you, it is not my intention).
Also, my user is a sudoers user, but still... it should have a solution. I also see that there is lot of Windows friendly code in the source code, that is also why i'm thinking (maybe i'm wrong but there is some clue) that we can hope to get a specific Linux user and Pythonist more knowledge-able to help to write a more compatible code for this platform or fix that point.
I don't said what you said is wrong, i still don't know but it doesn't convince me for the reasons i got you.

Also, if you know an other one package more Linux friendly with USB, i would be happy to use it instead of loosing time to turn around for nothing.

Please, maybe you can accept to try to let it as an issue if my arguments (who are logic) can convince you but much more to try to probably get someone able to add some Linux friendly code to your Windows friendly one ? I think it should be a constructive and open wide decision that can help anyone else.

**Comment by jonasmalacofilho** [comment]
Please try to consider that I might know a bit about what I'm talking about, at least in this instance.

Like I said: it's just how USB access works on Linux, plus what our API chooses to provide (the descriptors directly from the device).  So it's not a bug.

That said, at least in principle, reading those three descriptors from udev, when that's available, might be a valid proposal.  But claiming that not having such a feature is a bug, or claiming that we're not really compatible with Linux, or most of the rest in this thread just dilutes whatever value that proposal might have.

Plus, you failed to address the drawbacks I already mentioned.  And it also doesn't seem like you know a lot about udev, so this is actually a meta-proposal: hey Jonas, look into udev and figure _if_ there's a proposal to be made for reading the string descriptors from udev.

---

> you failed to demonstrate that it is not a bug

I could argue that the burden of proof has to be on who's proposing that a bug exists; otherwise you're making me prove to you that there is no bug, which can be substantially harder (to prove and to understand).

But, in the end, it doesn't even matter: if you want _me_ to fix a bug, you need to convince _me_ that it is a bug.  And you failed to do that.

(Probably because, to be fair, it's honestly really hard to claim that something that works and is used by lots of projects [including Linux ones] is broken in a such a fundamental way.  It just doesn't work how you expect it to, and even then it works just like Linux by default wants it to work: no accessing USB devices without special permissions).

By the way, even if you do get someone else to "fix it", I'm still the one that has to maintain the "fix", so you would still need to convince me that all drawbacks would be worth it.

> get a specific Linux user

You mean _another_ one, as I already mentioned to primarily run Linux.

> and Pythonist more knowledge-able

Ouch.

(It's true though: there are obviously people that are more knowledgeable than me in Linux and/or Python).

> some Linux friendly code to your Windows friendly one

The fact that there is "Windows-friendly" code in this project doesn't mean that we're Linux-unfriendly.

In reality, it only means that, from our point of view, Windows has more "quirks" for us to deal with.  Which, if you think about it, kinda shows that if we're at fault here, is from being Linux-centric.

> Please, maybe you can accept to try to let it as an issue [...]

Your arguments are already presented and recorded here, there's no need for another thread.

**Comment by jerome-diver** [comment]
Please, when i read you, it looks like it is very difficult to not hurt your susceptibility. Consider that i'm not the best in English (who is not my language) to write the best diplomatic way (if there is one possible) to not touch your susceptibility. I'm very sorry for that.
There is no problem that a lot of people are better than us in Python (so don't feel offended that i hope to get one in touch to fix the bug), and i never said that you are anti-Linux or pro anything else... i just read the source code and see it is a Windows first friendly code. Isn't it really ? Can we have to deny facts that source code is showing now ?

> In reality, it only means that, from our point of view, Windows has more "quirks" for us to deal with. Which, if you think about it, kinda shows that if we're at fault here, is from being Linux-centric.

Read yourself before to post, please... there is no "centric" bug, just missing some functionality to rich the target in all platform regardless the great Readme.md file description that looks like a commercial fantastic description.

It need a visibility to get people in touch... why are you doing that (to remove the issue and deny it is an functionality issue) ? Please, try to be a bit more humble and accept constructive criticism.

Your code is beautiful and you are fantastic, wonderful... even if it is not PEP8 convention friendly too, it is nice, but open your mind, please.

**Comment by jonasmalacofilho** [comment, marked as answer]
To avoid further deviating from the main subject of this thread, I'll keep this especially direct and to the point.

I'm sorry, but I'm rejecting your feature request, based on how it introduces complexity and changes the specification of some of our APIs, for a gain that is way too small: read/write access to the device will be necessary anyway for all other operations.

I may revisit this decision in the future; no further action from you is necessary.  In case this decision is revisited, it should also be considered whether the feature shouldn't live higher in the call chain (i.e. in the backend libraries).
 
Thank you for your suggestion.

**Comment by jerome-diver** [comment]
Thank you for your open mind ability, your humility and the way you consider other to be able to help in your project. It is fantastic.
